[scrollbar-styling] Implement dynamic changing of scrollbar-width

nmoucht opened the pull request:
#### 56109dd1e516a76e825afb71ffdb0966ad962eb4
<pre>
[scrollbar-styling] Implement dynamic changing of scrollbar-width
<a href="https://bugs.webkit.org/show_bug.cgi?id=276520">https://bugs.webkit.org/show_bug.cgi?id=276520</a>
<a href="https://rdar.apple.com/131580592">rdar://131580592</a>

Reviewed by Simon Fraser.

Add code in styleDidChange to dynamically update the NSScrollerImp when the value of
scrollbar-width is changed. Also send this value through to the UIProcess to ensure this works
with UI-side compositing.

* LayoutTests/TestExpectations:
* Source/WebCore/platform/Scrollbar.cpp:
(WebCore::Scrollbar::updateScrollbarThickness):
* Source/WebCore/platform/Scrollbar.h:
(WebCore::Scrollbar::setWidthStyle):
* Source/WebCore/platform/ScrollbarsController.cpp:
(WebCore::ScrollbarsController::updateScrollbarsThickness):
* Source/WebCore/platform/mac/ScrollbarsControllerMac.h:
* Source/WebCore/platform/mac/ScrollbarsControllerMac.mm:
(WebCore::ScrollbarsControllerMac::updateScrollerStyle):
(WebCore::ScrollbarsControllerMac::scrollbarWidthChanged):
* Source/WebCore/platform/mock/ScrollbarThemeMock.cpp:
(WebCore::ScrollbarThemeMock::scrollbarThickness):
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::styleDidChange):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollbarsController.mm:
(WebKit::RemoteScrollbarsController::scrollbarWidthChanged):
(WebKit::RemoteScrollbarsController::updateScrollbarStyle):

Canonical link: <a href="https://commits.webkit.org/282018@main">https://commits.webkit.org/282018@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/99d0b5badbd16e24d5a1057e4dc2cca435636887

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/61256 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/40617 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/13837 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/65206 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/11805 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/63386 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/48295 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/12080 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/49507 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/8207 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/63290 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/37771 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/53075 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/30339 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/34449 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/10300 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/10718 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/56266 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/10595 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/66937 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/5203 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/10393 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/56881 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/5226 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/53038 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/57083 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13744 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/4306 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/36421 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/37504 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/38598 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/37248 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -344,8 +344,13 @@ void RenderBox::styleDidChange(StyleDifference diff, const RenderStyle* oldStyle
+        if (isDocElementRenderer)
+            view().frameView().scrollbarsController().scrollbarWidthChanged(newStyle.scrollbarWidth());
+        else if (auto* scrollableArea = layer()->scrollableArea())
             scrollableArea->scrollbarsController().scrollbarWidthChanged(newStyle.scrollbarWidth());
```
`ScrollableArea` has a virtual `scrollbarWidthChanged()` so I think it would be cleaner for the ScrollableArea to talk to its scrollbarsController.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/platform/mac/ScrollbarsControllerMac.mm:
```diff
@@ -1029,6 +1029,19 @@ - (void)invalidate
 
+void ScrollbarsControllerMac::scrollbarWidthChanged(WebCore::ScrollbarWidth)
+{
+    auto needsScrollerStyleUpdate = m_needsScrollerStyleUpdate;
```
The cool kids are doing `auto needsScrollerStyleUpdate = std::exchange(m_needsScrollerStyleUpdate, false)`

smfr replied:
Actually this is more like a `SetForScope<>`. But why do we reset to false? A comment would help.

nmoucht replied:
This was originally to force a layout here, but I think it's actually not necessary. Removed.

nmoucht committed:
[scrollbar-styling] Implement dynamic changing of scrollbar-width
https://bugs.webkit.org/show_bug.cgi?id=276520
rdar://131580592

Reviewed by Simon Fraser.

Add code in styleDidChange to dynamically update the NSScrollerImp when the value of
scrollbar-width is changed. Also send this value through to the UIProcess to ensure this works
with UI-side compositing.

* LayoutTests/TestExpectations:
* Source/WebCore/platform/Scrollbar.cpp:
(WebCore::Scrollbar::updateScrollbarThickness):
* Source/WebCore/platform/Scrollbar.h:
(WebCore::Scrollbar::setWidthStyle):
* Source/WebCore/platform/ScrollbarsController.cpp:
(WebCore::ScrollbarsController::updateScrollbarsThickness):
* Source/WebCore/platform/mac/ScrollbarsControllerMac.h:
* Source/WebCore/platform/mac/ScrollbarsControllerMac.mm:
(WebCore::ScrollbarsControllerMac::updateScrollerStyle):
(WebCore::ScrollbarsControllerMac::scrollbarWidthChanged):
* Source/WebCore/platform/mock/ScrollbarThemeMock.cpp:
(WebCore::ScrollbarThemeMock::scrollbarThickness):
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::styleDidChange):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollbarsController.mm:
(WebKit::RemoteScrollbarsController::scrollbarWidthChanged):
(WebKit::RemoteScrollbarsController::updateScrollbarStyle):

Canonical link: https://commits.webkit.org/282018@main
